cupy/cuda/cupy_cub.cu

```diff
@@ -63,6 +63,8 @@ template <> struct NumericTraits<complex<double>> : BaseTraits<FLOATING_POINT, t
 // TODO(leofang): wrap CuPy's thrust namespace with another one (say, cupy::thrust) for safer scope resolution?
 
 namespace std {
+// TODO(amathews-amd): Check if this code can be safetly removed.
+#if HIP_VERSION < 401
```

Comment:
Could you install pytest and run this test locally to see it'd fail?
```
pytest tests/cupy_tests/core_tests/test_ndarray_reduction.py -k TestCubReduction
```

Reply:
Fails with the change on ROCm4.1 (See below) . Any ideas on how to fix ? 

`
cupy_tests/core_tests/test_ndarray_reduction.py ..F...F...F...F...F...F...F...F................................. 

FAILED cupy_tests/core_tests/test_ndarray_reduction.py::TestCubReduction_param_0_{backend='device', order='C', shape=(10,)}::test_cub_min - AssertionError:
FAILED cupy_tests/core_tests/test_ndarray_reduction.py::TestCubReduction_param_1_{backend='device', order='C', shape=(10, 20)}::test_cub_min - AssertionError:
FAILED cupy_tests/core_tests/test_ndarray_reduction.py::TestCubReduction_param_2_{backend='device', order='C', shape=(10, 20, 30)}::test_cub_min - AssertionError:
FAILED cupy_tests/core_tests/test_ndarray_reduction.py::TestCubReduction_param_3_{backend='device', order='C', shape=(10, 20, 30, 40)}::test_cub_min - AssertionError:
FAILED cupy_tests/core_tests/test_ndarray_reduction.py::TestCubReduction_param_4_{backend='device', order='F', shape=(10,)}::test_cub_min - AssertionError:
FAILED cupy_tests/core_tests/test_ndarray_reduction.py::TestCubReduction_param_5_{backend='device', order='F', shape=(10, 20)}::test_cub_min - AssertionError:
FAILED cupy_tests/core_tests/test_ndarray_reduction.py::TestCubReduction_param_6_{backend='device', order='F', shape=(10, 20, 30)}::test_cub_min - AssertionError:
FAILED cupy_tests/core_tests/test_ndarray_reduction.py::TestCubReduction_param_7_{backend='device', order='F', shape=(10, 20, 30, 40)}::test_cub_min - AssertionError:
_______________________
        numpy.testing.assert_allclose(
            cupy.asnumpy(actual), cupy.asnumpy(desired),
>           rtol=rtol, atol=atol, err_msg=err_msg, verbose=verbose)
E       AssertionError:
E       Not equal to tolerance rtol=1e-05, atol=0
E
E       Mismatched elements: 1 / 1 (100%)
E       Max absolute difference: 9.507621
E       Max relative difference: 1.
E        x: array(0.+0.j, dtype=complex64)
E        y: array(3.834415+8.700122j, dtype=complex64)
`

Reply:
@amathews-amd this is why we needed these lines...😓

This is not a proper fix. It breaks NumPy's [lexical comparison rules for complex numbers](https://numpy.org/doc/stable/reference/generated/numpy.sort.html):
> The sort order for complex numbers is lexicographic. If both the real and imaginary parts are non-nan then the order is determined by the real parts except when they are equal, in which case the order is determined by the imaginary parts.

So you can see that all `max`/`min` tests are failing with complex numbers. (I think in this test we test all dtypes as I figured out a way to make CUB work for everything we do!)

No one is happy with the way NumPy handles the comparison of complex numbers, but the upstream effort to change it is stalled, so unfortunately before it lands in NumPy's main branch we must comply...

The root cause of the build failure is reported in https://github.com/ROCm-Developer-Tools/HIP/issues/2247. Basically the bug is introduced in ROCm 4.1.0 and **its impact is beyond the build error here**. Now hip-clang forbids any possibility of specializing `std::numeric_limits<T>` for non-standard `T`, which IMHO is a *series bug against the C++ standard*. The bug itself should really be fixed asap upstream...

I would suggest you to looking into workarounds for us to 1. build successfully and 2. stay at the same level of the NumPy compliance that we have strived very hard to achieve. I don't think I know enough the internal of rocPRIM so this approach was the only way I found 😅 but it has been working very well since ROCm 3.5.0...

Reply:
> 2. stay at the same level of the NumPy compliance that we have strived very hard to achieve.

It is up to the team to decide if we would allow an exception to be made. For example, we could disable the code path for complex numbers on ROCm 4.1.0 if there is really no other way to work around...

Reply:
>Now hip-clang forbids any possibility of specializing std::numeric_limits<T> for non-standard T, which IMHO is a serious bug violating the C++ standard. The bug itself should really be fixed asap upstream...

cc @yxsamliu, would you help review the reported bug and share you insights if any? Thanks..

Reply:
@sunway513, thanks, but @yxsamliu already did, see https://github.com/ROCm-Developer-Tools/HIP/issues/2247#issuecomment-808507106.

Reply:
@leofang aha, thanks for the heads up and glad it's fully triaged! We'll follow up internally for resolution..

Reply:
Thank you @sunway513 (and @yxsamliu for tracking down the problem)! Please keep us posted if you have a decision 🙂

Reply:
Adding comment for note:
Solved in trunk: https://github.com/ROCm-Developer-Tools/HIP/issues/2247#issuecomment-816675887